File: OpenMP/C/01 Loop Decompose.c
#include<stdio.h>
#include<stdlib.h>
#include<omp.h>

#define MAX_ITS 10000

int main(int argc, char **argv)
{

  int nproc, thread_id, i, it_total;
  int *its_per_proc;

  nproc = omp_get_max_threads();
  its_per_proc = calloc(nproc, sizeof(int));
#pragma omp parallel for
  for (i = 0; i< MAX_ITS; ++i){
    thread_id = omp_get_thread_num();
    its_per_proc[thread_id] ++;
  }
/*No end here like in Fortran. Only for parallel for*/

  it_total = 0;
  for (i = 0; i< nproc; ++i){
    printf("Processor %i performed %i iterations\n", i, its_per_proc[i]);
    it_total += its_per_proc[i];
  }

  printf("Total work on all processors is %i\n", it_total);
}

File: OpenMP/Fortran/01 Loop Decompose.f90
PROGRAM loop_decompose

  USE omp_lib
  IMPLICIT NONE
  INTEGER, PARAMETER :: max_its = 10000
  INTEGER :: nproc, i, thread_id
  INTEGER, DIMENSION(:), ALLOCATABLE :: its_per_proc

  nproc = omp_get_max_threads()
  ALLOCATE(its_per_proc(nproc))
  its_per_proc = 0

!$OMP PARALLEL DO
  DO i = 1, max_its
    !This gets a unique ID number for each thread running from 0 -> nproc-1
    !You shouldn't really keep checking this value inside the loop because
    !it won't change but it makes the parallelism a bit simpler
    thread_id = omp_get_thread_num() + 1
    its_per_proc(thread_id) = its_per_proc(thread_id) + 1
  END DO
!$OMP END PARALLEL DO
!After the END PARALLEL DO construct the threads all connect back together
!and there is only one processor after this point

  DO i = 1, nproc
    PRINT '(A, I0, A, I0, A)', 'Processor ', i, ' performed ', &
        its_per_proc(i), ' iterations'
  END DO

  PRINT '(A, I0)','Total work on all processors is ', SUM(its_per_proc)

END PROGRAM loop_decompose
